### ISSUE TITLE
Simplify the input memory refinement of calls if there is no store

### ISSUE THREAD

**Issue by aqjune** [AliveToolkit/alive2#598]
If there is no store in a program, encoding the input memory refinement of function calls isn't necessary.
I found that this pattern happens frequently from LLVM unit tests (e.g. Transforms/InstCombine/partally-redundant-left-shift-input-masking-variant-a.ll ).
An example:
```
<src>
call f(poison)
call f(2)
<tgt>
call f(1)
call f(2)
```

Since the initial memory is equivalent, the first function call pair(`f(poison)` - `f(1)`) does not need to consider the input memory refinement.
After the execution of the first function call pair, the refinement between output memory is guaranteed.
Since there is no store between calls, the input memory refinement of the second call pair (`f(2)` in src and tgt) trivially holds.

**Comment by nunoplopes** [comment]
What's the issue here? Is it because memory refinement is being called?
Because it should be folding to true straight away right? Since f(1) only refines the first call of src, it get the same memory. In the future it may get a refined memory, but for now they are equal.
Or am I missing smth?

**Comment by aqjune** [comment]
Yes, it is about optimizing the SMT query (not a correctness issue).

**Comment by nunoplopes** [comment]
I understand it's not a correctness issue.
My question is whether there's anything to optimize? Isn't the SMT query simply false?
Is it what you are proposing to skip vcgen for this case?

**Comment by aqjune** [comment]
The example was simply raising false, sorry. The fn call optimization using ranges and fn call reuses in src was effective enough to fold them; I should have checked that. :(

But, by slightly changing the optimization I think this example can gain benefit: [identity.ll.zip](https://github.com/AliveToolkit/alive2/files/5615949/identity.ll.zip)


The domain of g() calls in tgt is big (I printed `domain` variable at `State::addFnCall`).
But, I think it can be optimized to simply true because for each g() in src, f() is preceding it.

Let's consider `g2` call in tgt.
`f2` in tgt did not raise UB when `g2` is reached, so it must have been one of f() calls in src, say `f1`.
Then, we can always match the following g() call (`g1` in this case) with `g2`.
Since we can always pick one g() call from src, the domain is always true.

**Comment by nunoplopes** [comment]
Ok, I see the issue. Here's a reduced test case:
```
declare void @f()

@x = global i32 0
define void @src(i1 %c) {
  br i1 %c, label %A, label %B
A:
  call void @f()
  br label %EXIT
B:
  store i32 0, i32* @x
  call void @f()
  br label %EXIT
EXIT:
  ret void
}

define void @tgt(i1 %c) {
  br i1 %c, label %A, label %B
A:
  call void @f()
  br label %EXIT
B:
  store i32 0, i32* @x
  call void @f()
  br label %EXIT
EXIT:
  ret void
}
```

The issue is that memory refinement doesn't simplify enough the query between the memory with/without the store. It should simplify to false, but instead it produces a complicated formula saying the memories are equal except for global @x. Which is not very interesting for function calls.
Let's attempt to fix this in the new model. (that's my Christmas present hopefully).
